File: src/java/MainOrdenamiento.java
import java.io.File;
import java.io.FileNotFoundException;
import java.io.PrintWriter;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Scanner;
import java.util.Map;
import java.util.HashMap;

public class MainOrdenamiento {

    public static int[] leerDatos(String archivo) throws FileNotFoundException {
        List<Integer> numeros = new ArrayList<>();
        Scanner scanner = new Scanner(new File(archivo));
        while (scanner.hasNextInt()) {
            numeros.add(scanner.nextInt());
        }
        scanner.close();
        
        int[] arr = new int[numeros.size()];
        for (int i = 0; i < numeros.size(); i++) {
            arr[i] = numeros.get(i);
        }
        return arr;
    }

    public static void main(String[] args) throws FileNotFoundException {
        System.err.println("Iniciando benchmarks de ORDENAMIENTO en Java...");
        
        String[] nombresAlgoritmos = {"ShakerSort", "DualPivotQuickSort", "HeapSort", "MergeSort", "RadixSort"};
        int[] tamaños = {10000, 100000, 1000000};
        
        Map<String, String> complejidades = new HashMap<>();
        complejidades.put("ShakerSort", "O(n²)");
        complejidades.put("DualPivotQuickSort", "O(n log n)");
        complejidades.put("HeapSort", "O(n log n)");
        complejidades.put("MergeSort", "O(n log n)");
        complejidades.put("RadixSort", "O(n·k)");
        
        try (PrintWriter writer = new PrintWriter("results/tiempos_ordenamiento_java.csv")) {
            
            writer.println("algoritmo;complejidad;lenguaje;tamaño;tiempo");
            
            for (int tam : tamaños) {
                String rutaArchivo = String.format("data/datos_%d.txt", tam);
                int[] datosOriginales = leerDatos(rutaArchivo);
                
                System.err.println(String.format("\n--- Probando con %d elementos ---", tam));

                for (String nombreAlg : nombresAlgoritmos) {
                    String complejidad = complejidades.get(nombreAlg);

                    if (nombreAlg.equals("ShakerSort") && tam == 1000000) {
                        writer.println(String.format("ShakerSort;%s;Java;1000000;", complejidad));
                        System.err.println(String.format("--- Omitiendo ShakerSort (%s) para 1,000,000 elementos por ser demasiado lento ---", complejidad));
                        continue;
                    }
                    
                    int[] datosACopiar = Arrays.copyOf(datosOriginales, datosOriginales.length);
                    
                    long startTime = System.nanoTime();

                    switch (nombreAlg) {
                        case "ShakerSort": Ordenamiento.shakerSort(datosACopiar); break;
                        case "DualPivotQuickSort": Ordenamiento.dualPivotQuickSort(datosACopiar, 0, datosACopiar.length - 1); break;
                        case "HeapSort": Ordenamiento.heapSort(datosACopiar); break;
                        case "MergeSort": Ordenamiento.mergeSort(datosACopiar, 0, datosACopiar.length - 1); break;
                        case "RadixSort": Ordenamiento.radixSort(datosACopiar); break;
                    }

                    long endTime = System.nanoTime();
                    double tiempoTotal = (endTime - startTime) / 1e9;
                    
                    System.err.println(String.format("  - %s (%s): %.4f segundos", nombreAlg, complejidad, tiempoTotal));
                    writer.println(String.format("%s;%s;Java;%d;%.4f", nombreAlg, complejidad, tam, tiempoTotal));
                }
            }
        } 

        System.err.println("\nBenchmarks en Java finalizados. Archivo CSV generado.");
    }
}

File: main_ordenamiento.py
import time
import pandas as pd
# Importamos todas las funciones que pegaste en el otro archivo
from src.python.ordenamiento import *

def leer_datos(archivo):
    """Carga los números desde un archivo de texto."""
    with open(archivo, 'r') as f:
        return [int(line.strip()) for line in f]

# --- Bucle principal para medir tiempos ---
print("Iniciando benchmarks de ORDENAMIENTO en Python...")

resultados = []
# Diccionario que asocia el nombre del algoritmo con su función
algoritmos = {
    "ShakerSort": (shaker_sort, "O(n²)"),
    "DualPivotQuickSort": (dual_pivot_quicksort, "O(n log n)"),
    "HeapSort": (heap_sort, "O(n log n)"),
    "MergeSort": (merge_sort, "O(n log n)"),
    "RadixSort": (radix_sort, "O(n·k)")
}

tamaños = [10000, 100000, 1000000]

for tam in tamaños:
    ruta_archivo = f"data/datos_{tam}.txt"
    datos = leer_datos(ruta_archivo)
    print(f"\n--- Probando con {tam} elementos ---")

    for nombre, (funcion_sort, complejidad) in algoritmos.items():

        if nombre == "ShakerSort" and tam == 1000000:
            print(f" {nombre} ({complejidad}): Omitido por ser demasiado lento (estimado > 10 horas)")
            # Añadimos un resultado simbólico para que el CSV quede completo
            resultados.append({
                "algoritmo": nombre,
                "complejidad": complejidad,
                "lenguaje": "Python",
                "tamaño": tam,
                "tiempo": float('inf') # Usamos infinito para representar el tiempo
            })
            continue
        # ¡Importante! Copiamos los datos para que cada algoritmo ordene el arreglo original
        datos_a_ordenar = datos.copy()

        start_time = time.perf_counter()
        funcion_sort(datos_a_ordenar)
        end_time = time.perf_counter()

        tiempo_total = end_time - start_time
        print(f" check : {nombre} ({complejidad}): {tiempo_total:.4f} segundos")
        
        resultados.append({
            "algoritmo": nombre,
            "complejidad": complejidad,
            "lenguaje": "Python",
            "tamaño": tam,
            "tiempo": tiempo_total
        })

# --- Guardar resultados en CSV ---
df_resultados = pd.DataFrame(resultados)
ruta_csv = "results/tiempos_ordenamiento_python.csv"
df_resultados.to_csv(ruta_csv, index=False)

print(f"\n¡Benchmarks finalizados! Resultados guardados en '{ruta_csv}'")